rotkehlchen/exchanges/kraken.py

```diff
@@ -782,3 +779,282 @@ def query_online_income_loss_expense(
+            timestamp = trade_parts[0].timestamp
+            base_trade, quote_trade, fee_part, kfee_part = None, None, None, None
+            for trade_part in trade_parts:
+                part_asset = trade_part.amount.asset
+                if trade_part.event_subtype == HistoryEventSubType.FEE:
+                    fee_part = trade_part
+                elif trade_part.amount.asset == A_KFEE:
+                    kfee_part = trade_part
+                elif part_asset == base_asset:
+                    base_trade = trade_part
+                elif part_asset == quote_asset:
+                    quote_trade = trade_part
+
+            if base_trade is None or quote_trade is None:
+                self.msg_aggregator.add_error(f'Failed to read trades. {trade_parts}')
+                continue
+
+            base_amount = base_trade.amount.balance.amount
+            quote_amount = quote_trade.amount.balance.amount
+
+            if quote_amount < 0:
+                trade_type = TradeType.SELL
+                amount = base_amount
+                received_asset = quote_trade.amount.asset
+            else:
+                trade_type = TradeType.BUY
+                amount = quote_amount
+                received_asset = base_trade.amount.asset
+
+            if base_amount != ZERO:
+                rate = Price(quote_amount / base_amount)
+            else:
+                raise ZeroDivisionError(
+                    f'Rate for trade couldnt be calculated. Base amount is ZERO '
+                    f'for {base_trade.event_identifier}',
+                )
+
+            exchange_uuid = (
+                str(base_trade.event_identifier) +
+                str(timestamp)
+            )
+            if fee_part is not None and kfee_part is not None:
+                fee = Fee(fee_part.amount.balance.amount)
+                fee_asset = fee_part.amount.asset
+                # Here we should create a ledger action for kfee
+            elif (None, None) == (fee_part, kfee_part):
+                fee = Fee(ZERO)
+                fee_asset = received_asset
+            elif fee_part is not None:
+                fee = Fee(fee_part.amount.balance.amount)
+                fee_asset = fee_part.amount.asset
+            elif kfee_part is not None:
+                fee = Fee(kfee_part.amount.balance.amount)
+                fee_asset = A_KFEE
+
+            trade = Trade(
+                timestamp=timestamp,
+                location=Location.KRAKEN,
+                base_asset=base_asset,
+                quote_asset=quote_asset,
+                trade_type=trade_type,
+                amount=AssetAmount(amount),
+                rate=rate,
+                fee=fee,
+                fee_currency=fee_asset,
+                link=exchange_uuid,
+            )
+            trades.append(trade)
+        return trades
+
+    def query_kraken_ledger_actions(self) -> None:
+        with self.ledger_lock:
```

Comment:
Just forget about the ledger_lock and use the `protect_with_lock()` modifier since all you do is put the entire function under a lock. This is exactly what we made that modifier for. And will make the code more readable.